producer.go

```diff
@@ -0,0 +1,343 @@
+		config.MaxConcurrency = DefaultMaxConcurrency
+	}
+
+	if config.DialTimeout == 0 {
+		config.DialTimeout = DefaultDialTimeout
+	}
+
+	if config.ReadTimeout == 0 {
+		config.ReadTimeout = DefaultReadTimeout
+	}
+
+	if config.WriteTimeout == 0 {
+		config.WriteTimeout = DefaultWriteTimeout
+	}
+
+	if config.MaxRetryTimeout == 0 {
+		config.MaxRetryTimeout = DefaultMaxRetryTimeout
+	}
+
+	if config.MinRetryTimeout == 0 {
+		config.MinRetryTimeout = DefaultMinRetryTimeout
+	}
+
+	p = &Producer{
+		reqs:            make(chan ProducerRequest, config.MaxConcurrency),
+		done:            make(chan struct{}),
+		address:         config.Address,
+		topic:           config.Topic,
+		dialTimeout:     config.DialTimeout,
+		readTimeout:     config.ReadTimeout,
+		writeTimeout:    config.WriteTimeout,
+		maxRetryTimeout: config.MaxRetryTimeout,
+		minRetryTimeout: config.MinRetryTimeout,
+	}
+	p.join.Add(config.MaxConcurrency)
+
+	for i := 0; i != config.MaxConcurrency; i++ {
+		go p.run()
+	}
+
+	return
+}
+
+// Stop gracefully shutsdown the producer, cancelling all inflight requests and
+// waiting for all backend connections to be closed.
+//
+// It is safe to call the method multiple times and from multiple goroutines,
+// they will all block until the producer has been completely shutdown.
+func (p *Producer) Stop() {
+	p.once.Do(p.stop)
+	err := errors.New("publishing to a producer that was already stopped")
+
+	for req := range p.reqs {
+		req.complete(err)
+	}
+
+	p.join.Wait()
+}
+
+// Publish sends a message using the producer p, returning an error if it was
+// already closed or if an error occurred while publishing the message.
+//
+// Note that no retry is done internally, the producer will fail after the
+// first unsuccessful attempt to publish the message. It is the responsibility
+// of the caller to retry if necessary.
+func (p *Producer) Publish(message []byte) (err error) {
+	defer func() {
+		if recover() != nil {
+			err = errors.New("publishing to a producer that was already stopped")
+		}
+	}()
+
+	res := make(chan error, 1)
+
+	p.reqs <- ProducerRequest{
+		Message:  message,
+		Response: res,
+	}
+
+	err = <-res
+	return
+}
+
+// Requests returns a write-only channel that can be used to submit requests to p.
+//
+// This method is useful when the publish operation needs to be associated with
+// other operations on channels in a select statement for example, or to publish
+// in a non-blocking fashion.
+func (p *Producer) Requests() chan<- ProducerRequest {
+	return p.reqs
+}
+
+func (p *Producer) stop() {
+	close(p.done)
+	close(p.reqs)
+}
+
+func (p *Producer) run() {
+	var conn *Conn
+	var pipe chan ProducerRequest
+	var ping chan struct{}
+	var retry time.Duration
+
+	defer p.join.Done()
+	defer func() {
+		if conn != nil {
+			conn.Close()
+		}
+		if pipe != nil {
+			close(pipe)
+		}
+	}()
+
+	for {
+		select {
+		case <-p.done:
+			return
+
+		case <-ping:
+			if err := p.ping(conn); err != nil {
+				close(pipe)
```

Comment:
wonder if we can move the cleanup step around, like:

```
cleanup := func(){
  close(pipe)
  ...
}
```

Just so that cleanup doesn't go out of sync.